File: 450-delete-node-in-a-bst/delete-node-in-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode inorderSuccessor(TreeNode root){
        while(root.left != null){
            root = root.left;
        }
        return root;
    }

    public TreeNode deleteNode(TreeNode root, int key) {
        if(root == null) {
            return null;
        }

        if(key > root.val) {
            root.right = deleteNode(root.right, key);
        } 
        else if(key < root.val) {
            root.left = deleteNode(root.left, key);
        }
        else if(key == root.val) {
            if(root.left==null && root.right==null){
                return null;
            }
            else if(root.left==null || root.right==null){
                TreeNode newRoot = root.left==null? root.right : root.left;
                return newRoot; 
            } 
            TreeNode IS = inorderSuccessor(root.right);
            root.val = IS.val;
            root.right = deleteNode(root.right, IS.val);
            return root;
        }
        return root;
    }
}

File: 450-delete-node-in-a-bst/delete-node-in-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

    TreeNode* inorderSuccessor(TreeNode* root) {
        while(root->left!=NULL) {
            root=root->left;
        }
        return root;
    }

    TreeNode* deleteNode(TreeNode* root, int key) {
        if(root == NULL) {
            return NULL;
        }

        if(key < root -> val) {
            root -> left = deleteNode(root->left, key);
        }

        if(key > root -> val) {
            root -> right = deleteNode(root->right, key);
        }

        if(key == root ->val) {
            if(root->left == NULL && root->right == NULL) {
                delete root;
                return NULL;
            }

            if (root->left == NULL) {
                TreeNode* node = root->right;
                delete root;
                return node;
            } else if (root->right == NULL) {
                TreeNode* node = root->left;
                delete root;
                return node;
            }

            TreeNode* IS = inorderSuccessor(root->right);
            root -> val = IS -> val;
            root -> right = deleteNode(root->right, IS->val);
            return root;
        }
        return root;
    }
};